Implements knight valid_move? method.

dmk1991 opened the pull request:
# valid_move? for knight pieces
Good morning everyone, PR to merge in method to check if a knight is making a valid move along with a suite of rspec tests to go with it. Method does the following,

* Calculate the distance from piece's current row and column to destination row and column.
* Returns true if move is a 2/1 ratio, false if not.

Please double check that my logic is sound and let me know of any other recommendations.

dmk1991 committed:
Implements knight valid_move? method.

jacobwgillespie commented on app/models/knight.rb:
```diff
@@ -1,2 +1,8 @@
+    row_diff = (self.row - row_destination).abs
+    col_diff = (self.column - col_destination).abs
+
+    (row_diff / col_diff) == 2 || (col_diff / row_diff) == 2 ? true : false
```
This looks like a good approach, however larger jumps will also trigger as the correct ratio (i.e. moving up 4 and over 2), and additionally Ruby drops precision on integer arithmetic, so any remainders of those divisions will get dropped (up 5 over 2 will still be 2):

```ruby
4 / 2 == 2 #=> true
5 / 2 == 2 #=> true
```

You'll likely have to compare all the values directly - you can actually compare an array to an array to make it simple:

```ruby
[1, 2] == [row_diff, col_diff] || #...
```

Also, one small style bit, if the results of an if are true and false, the if can be removed entirely and the condition can be used directly:

```ruby
def verbose_one
  if row_diff == 2
    true
  else
    false
  end
end

def verbose_two
  row_diff == 2 ? true : false
end

def simple
  row_diff == 2
end
```

```ruby
def other_verbose
  result = something == 2 ? true : false
  # ...
end

def other_simple
  result = something == 2
  # ...
end
```

Looking good!  🎉

dmk1991 commented:
@jacobwgillespie Awesome, thanks for the good point about bigger jumps. When everything worked on the first try I was a little skeptical about it's correctness haha.  Also, very good to know about how Ruby handles arithmetic, I was not aware of that. I will implement these changes and push asap.

jacobwgillespie commented:
No prob!  The arithmetic bit is a carryover from how the CPU actually does the math - you have the same kind of behavior with a lower-level language like C.  If you're dividing an integer by an integer, you will get an integer result - if you want to do floating point arithmetic, you have to convert to floats:

```ruby
5.to_f / 2.to_f #=> 2.5

5.0 / 2.0 #=> 2.5
```

With ruby, it's smart enough so that if one of the values is a float, it will have a float result:

```ruby
5 / 2.0 #=> 2.5
```

With a statically typed language like C, they have to be the same type (originally or converted).

dmk1991 committed:
Fixes valid_move to only allow 2/1 moves

dmk1991 commented:
Implemented the recommended change from Jacob and added another test to make sure the piece is only moving one jump one 2/1, not multiples (i.e. 4/2, 6/3).

Thanks for the lesson Jacob, that's cool to know that this functionality within languages comes from how the actual CPU handles the task.

jacobwgillespie reviewed (approved):
LGTM!
